core/src/utils/focus-trap.ts

```diff
@@ -0,0 +1,82 @@
+import { focusVisibleElement } from '@utils/helpers';
+
+/**
+ * This query string selects elements that
+ * are eligible to receive focus. We select
+ * interactive elements that meet the following
+ * criteria:
+ * 1. Element does not have a negative tabindex
+ * 2. Element does not have `hidden`
+ * 3. Element does not have `disabled` for non-Ionic components.
+ * 4. Element does not have `disabled` or `disabled="true"` for Ionic components.
+ * Note: We need this distinction because `disabled="false"` is
+ * valid usage for the disabled property on ion-button.
+ */
+export const focusableQueryString =
+  '[tabindex]:not([tabindex^="-"]):not([hidden]):not([disabled]), input:not([type=hidden]):not([tabindex^="-"]):not([hidden]):not([disabled]), textarea:not([tabindex^="-"]):not([hidden]):not([disabled]), button:not([tabindex^="-"]):not([hidden]):not([disabled]), select:not([tabindex^="-"]):not([hidden]):not([disabled]), .ion-focusable:not([tabindex^="-"]):not([hidden]):not([disabled]), .ion-focusable[disabled="false"]:not([tabindex^="-"]):not([hidden])';
+
+/**
+ * Focuses the first descendant in a context
+ * that can receive focus. If none exists,
+ * a fallback element will be focused.
+ * This fallback is typically an ancestor
+ * container such as a menu or overlay so focus does not
+ * leave the container we are trying to trap focus in.
+ */
+export const focusFirstDescendant = <T extends HTMLElement>(ref: Element, fallbackElement: T) => {
```

Comment:
Separate comment, but on the same function. Why is the fallback argument required if in most of the implementations we supply the same element reference to both the `ref` and `fallbackElement`? 

Could we make it so that if you pass a `ref` without a `fallbackElement`, it uses the `ref` as the fallback?

Reply:
Updated in https://github.com/ionic-team/ionic-framework/pull/29091/commits/b1917462ab61c9e73986e190c7a903ecf044ffad